keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,919 @@
+- Support for certificate based signing via the `x5c` header field (this is deferred until we have more user evidence)
+- Supporting access to the JWT header in rules (this is deferred until we have more user evidence)
+- Supporting access to the provider config itself in rules (this is deferred until we have more user evidence)
+- Supporting JWTs with multiple signatures (to avoid any security issues caused by signature confusion)
+
+## Proposal
+
+1. Add new `apiserver.config.k8s.io` API object to parse a structured config file `AuthenticationConfiguration`.
+2. Add a single flag `--authentication-config` for kube-apiserver to point to the structured config file, automatically reload it on changes.
+3. Use an expression language to let users write their own logic for mappings and validation rules
+  (expressions should be simple for common cases, yet powerful to cover most user stories).
+
+### Risks and Mitigations
+
+Since this is a new optional feature, no migration is required unless the user wants to replace their
+existing OIDC flags usage with the config file.  The use of `--authentication-config` is mutually exclusive
+with the existing OIDC flags, so we will provide documentation for migrating from a flag-based config to the new config.
+
+## Design Details
+
+### Configuration file
+
+TODO:
+
+- should we have any revocation mechanism?
+  => use revocation endpoint if it is in the discovery document? (lets decide what we want here before beta)
+  => related issue https://github.com/kubernetes/kubernetes/issues/71151
+- distributed claims with fancier resolution requirements (such as access tokens as input)
+- implementation detail: we should probably parse the `iss` claim out once
+- should audit annotations be set on validation failure?
+- should we introduce implicit checks that assure at least one of uid/username are nonempty?
+- decide what error should be returned if CEL eval fails at runtime
+  `500 Internal Sever Error` seem appropriate but authentication can only do `401`
+
+The main part of this proposal is a configuration file. It contains an array of providers:
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1alpha1
+kind: AuthenticationConfiguration
+jwt:
+- issuer:
+    url: https://example.com
+    clientIDs:
+    - my-app
+  claimValidationRules:
+  - claim: hd
+    requiredValue: example.com
+  - expression: 'claims.hd == "example.com"'
+    message: the hd claim must be set to example.com
+  - expression: 'claims.exp - claims.nbf <= 86400'
+    message: total token lifetime must not exceed 24 hours
+  claimMappings:
+    username:
+      expression: 'claims.username + ":external-user"'
+    groups:
+      expression: 'claims.roles.split(",")'
+    uid:
+      claim: 'sub'
+    extra:
+    - key:
+        constant: 'client_name'  # TODO: decide if we really need this flexibility or can we just have constant keys
+      value:
+        claim: 'aud'
+  # TODO(enj): drop this and figure out to get from CEL
```

Comment:
let's drop the TODO from the API portion and switch this to a section discussing how to resolve distributed claims used from CEL

```yaml
claims = {
  "foo":"bar",
  "foo.bar": "...",
  "true": "...",
  "_claim_names": {
    "groups": "group_source"
   },
   "_claim_sources": {
    "group_source": {"endpoint": "https://example.com/claim_source"}
   }
}
```

capturing answers to questions during review:

1. what about claim names containing `.` or keywords? can reference using square-brackets:
  `claims['foo.bar']`
  `claims['true']`

2. for distributed claims, can we implement a CEL type resolver so that a cel expression `claims.foo` gets resolved via a distributed claim the first time it is used? yes, this seems likely and preferable so we only resolve the things we need (in case an early validation rule fails and short-circuits).

Reply:
Fixed.